src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.script.mustache;
+
+import com.github.mustachejava.DefaultMustacheFactory;
+import com.github.mustachejava.MustacheException;
+import com.github.mustachejava.reflect.ReflectionObjectHandler;
+import com.google.common.collect.Iterables;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.lang.reflect.Array;
+import java.util.*;
+
+/**
+ * A custom MustacheFactory that does simple JSON escaping and supports array/collection access syntax in the form
+ * of {@code array_or_collection.X} where {@code array_or_collection} refers to an array or a collection and
+ * {@code X} is the index into it. For example, referring to the "key" field in the 3rd item in the "array" array
+ * can be expressed as {@code array.3.key}.
+ */
+public final class InternalMustacheFactory extends DefaultMustacheFactory {
+
+    public InternalMustacheFactory() {
+        setObjectHandler(new InternalObjectHandler());
+    }
+
+    @Override
+    public void encode(String value, Writer writer) {
+        try {
+            escape(value, writer);
+        } catch (IOException e) {
+            throw new MustacheException("Failed to encode value: " + value);
+        }
+    }
+
+    public static Writer escape(String value, Writer writer) throws IOException {
```

Comment:
never looked into it... but I think that'll belong in a different PR... don't want to potentially break the escape functionality for the feature added here